New advection fields

sophieaerdker opened the pull request:
This PR adds new shock profiles for simulating diffusive shock acceleration to the AdvectionField module. In the 2023 ICRC proceedings the advection field `OneDimensionalAdvectionShock` was used to obtain the time-dependent spectra at a 1D cartesian shock. Additional shock profiles allow for oblique shocks and 1D radial shocks.

JulienDoerner commented on src/advectionField/AdvectionField.cpp:
```diff
@@ -177,6 +177,217 @@ std::string SphericalAdvectionField::getDescription() const {
+
+	if (cool_ups == true){
+		if (r <= r_sh)
+			return 2*a/r - 2*b/r * tanh((r-r_sh)/l_sh) - b/l_sh * (1 - tanh((r-r_sh)/l_sh)*tanh((r-r_sh)/l_sh));
```
you often use the term `tanh((r - r_sh)  /  l_sh)`. I would precalculate it und use it as a variable to speed up the calculation a bit.

JulienDoerner commented on src/advectionField/AdvectionField.cpp:
```diff
@@ -177,6 +177,217 @@ std::string SphericalAdvectionField::getDescription() const {
+
+std::string OneDimensionalSphericalShock::getDescription() const {
+	std::stringstream s;
+	s << "Shock width: " << l_sh / km  << " km, ";
```
add module name

JulienDoerner commented on src/advectionField/AdvectionField.cpp:
```diff
@@ -177,6 +177,217 @@ std::string SphericalAdvectionField::getDescription() const {
+//----------------------------------------------------------------
+
+ObliqueAdvectionShock::ObliqueAdvectionShock(double r_comp, double vx_up, double vy, double x_sh){
+
```
whitespace

JulienDoerner commented on src/advectionField/AdvectionField.cpp:
```diff
@@ -177,6 +177,217 @@ std::string SphericalAdvectionField::getDescription() const {
+	}
+
+Vector3d ObliqueAdvectionShock::getField(const Vector3d &position) const {
+
```
whitespace

JulienDoerner commented on src/advectionField/AdvectionField.cpp:
```diff
@@ -177,6 +177,217 @@ std::string SphericalAdvectionField::getDescription() const {
+
+//----------------------------------------------------------------
+
+ObliqueAdvectionShock::ObliqueAdvectionShock(double r_comp, double vx_up, double vy, double x_sh){
```
space between `) {`

JulienDoerner commented on src/advectionField/AdvectionField.cpp:
```diff
@@ -177,6 +177,217 @@ std::string SphericalAdvectionField::getDescription() const {
+	double b = (vx_up - vx_down)*0.5;
+
+	Vector3d v(0.);
+    v.x = a - b*tanh(x/x_sh);
```
fix indenting

JulienDoerner commented on src/advectionField/AdvectionField.cpp:
```diff
@@ -177,6 +177,217 @@ std::string SphericalAdvectionField::getDescription() const {
+
+std::string ObliqueAdvectionShock::getDescription() const {
+	std::stringstream s;
+	s << "Shock width: " << x_sh / km  << " km, ";
```
module name

JulienDoerner commented:
The new module has some set-function for the parameters. For consistency also get-functions would be good. 

And for the functionality it would be good if the position of the shocks could also be shifted, to allow more customized simulations.

sophieaerdker commented on include/crpropa/advectionField/AdvectionField.h:
```diff
@@ -128,6 +128,96 @@ class SphericalAdvectionField: public AdvectionField {
+
+/**
+ @class 1DSphericalShock
+ @brief Advection field in r-direction with shock at r_sh and width l_sh approximated by tanh() 
```
It's called `OneDimensionalCartesianShock` now

sophieaerdker commented on src/advectionField/AdvectionField.cpp:
```diff
@@ -177,6 +177,217 @@ std::string SphericalAdvectionField::getDescription() const {
+}
+
+std::string OneDimensionalAdvectionShock::getDescription() const {
+	std::stringstream s;
```
This way, it's consistent with the other advection fields

sophieaerdker commented:
I added the get-functions and changed the name of `OneDimensionalAdvectionShock` to `OneDimensionalCartesianShock` to set apart from `OneDimensionalSphericalShock`. 

I think the change of the shock position is not necessary for the 1D cartesian case, since it's only for testing purposes and basically all other (analytical) works have the 1D shock at x=0. You can simply set your environment around it and it's not suitable for more complex scenarios anyways.

JulienDoerner commented on include/crpropa/advectionField/AdvectionField.h:
```diff
@@ -128,6 +128,112 @@ class SphericalAdvectionField: public AdvectionField {
+	double getVup() const;
+	double getShockwidth() const;
+	double getShockRadius() const;
+	double getCooling() const;
```
I guess this get-function should be a bool?

JulienDoerner commented on include/crpropa/advectionField/AdvectionField.h:
```diff
@@ -128,6 +128,112 @@ class SphericalAdvectionField: public AdvectionField {
+public:
+/** Constructor
+	@param compressionRatio //compression ratio of shock
+	@param vUpstream //upstream velocity 
```
This description is not the same variable name as in the constructer. This may lead to problems with the doxygen generation.

JulienDoerner commented on include/crpropa/advectionField/AdvectionField.h:
```diff
@@ -150,6 +256,7 @@ class SphericalAdvectionShock: public AdvectionField {
 	@param r_0 	Position of the shock
 	@param v_0 	Constant velocity (r<<r_o)
 	@param lambda 	Transition width / width of the shock
+	@param v_phi 	Rotation speed at r_rot
```
This is not a parameter of this constuctor
